src/pretty/list.rs

```diff
@@ -0,0 +1,276 @@
+        });
+
+        pretty_commented_items(
+            &self.items,
+            BoxDoc::text(","),
+            BoxDoc::text(","),
+            (BoxDoc::text("("), BoxDoc::text(")")),
+            false,
+            only_one_pattern,
+            false,
+            self.fold_style,
+        )
+    }
+
+    pub fn convert_params(mut self, params: Params<'a>, is_unnamed: bool) -> BoxDoc<'a, ()> {
+        self.process_list(params.to_untyped(), |node| self.printer.convert_param(node));
+
+        pretty_commented_items(
+            &self.items,
+            BoxDoc::text(","),
+            BoxDoc::text(","),
+            (BoxDoc::text("("), BoxDoc::text(")")),
+            false,
+            false,
+            is_unnamed
+                && is_exact_one_and(params.children(), |it| {
+                    matches!(
+                        *it,
+                        Param::Pos(Pattern::Normal(_)) | Param::Pos(Pattern::Placeholder(_))
+                    )
+                }),
+            self.fold_style,
+        )
+    }
+
+    fn process_list<T: AstNode<'a>>(
+        &mut self,
+        list_node: &'a SyntaxNode,
+        item_converter: impl Fn(T) -> BoxDoc<'a, ()>,
+    ) {
+        // Each item can be attached with a comment at the front and back.
+        // Can break line after the front attachment.
+        // If the back attachment appears before the comma, the comma is move to its front.
+        // Only one can attach after.
+
+        self.fold_style = self.printer.get_fold_style_ut(list_node);
+
+        for node in list_node.children() {
+            if let Some(item) = node.cast() {
+                self.item_count += 1;
+                let before = if self.free_comments.is_empty() {
+                    BoxDoc::nil()
+                } else {
+                    BoxDoc::intersperse(self.free_comments.drain(..), BoxDoc::line())
+                        .append(BoxDoc::line())
+                };
+                self.items.push(CommentedItem::Commented(
+                    before.append(item_converter(item)).group(),
+                    BoxDoc::nil(),
+                ));
+                self.can_attach = true;
+            } else if is_comment_node(node) {
+                // Line comment cannot appear in single line block
+                if node.kind() == SyntaxKind::LineComment {
+                    self.fold_style = FoldStyle::Never;
+                }
+                self.free_comments.push(comment(node));
+                // can_attach = false;
+            } else if node.kind() == SyntaxKind::Comma {
+                self.try_attach();
+            } else if node.kind() == SyntaxKind::Space {
+                let newline_cnt = node.text().chars().filter(|c| *c == '\n').count();
+                if newline_cnt > 0 {
+                    self.attach_or_detach();
+                    self.can_attach = false;
+                }
+            }
+        }
+        self.attach_or_detach();
```

Comment:
This is not a good name. It sounds like it would return a bool and it takes `&self`. But instead it returns nothing and mutates `self`